pkg/client/cache/fifo.go

```diff
@@ -22,12 +22,15 @@ import (
 	"k8s.io/kubernetes/pkg/util/sets"
 )
 
+type PopProcessFunc func(interface{}) error
+
 // Queue is exactly like a Store, but has a Pop() method too.
 type Queue interface {
 	Store
 
-	// Pop blocks until it has something to return.
-	Pop() interface{}
+	// Pop blocks until it has something to process.
+	// It returns the object that was process and the result of processing.
+	Pop(PopProcessFunc) (interface{}, error)
```

Comment:
@wojtek-t 

I understand the intention. However, as far as I can see we do not really need to change a lot of places if we try to approach this problem in a different way that solving it outside the queue. I may be wrong. And I am OK with this solution as a temp one.

So here is my thought:
We now try to force the dequeue operation and the processing of the dequeued item protected by a shared queue lock. To be more concrete, for the service controller it is its cache.

First, I do not believe this is not a good abstraction. Queue now needs to worry about the synchronization between internal and external objects.  Second, this does not prevent the synchronization problem to happen. The knowobject cache still can be accessed externally asynchronously. This solves today's issue in a not elegant way I think. 

Thus I dislike it. And I suggest to clean this whole thing up after 1.3 if we want to go this approach right now.